Implementation of application tests using configuration files

GernotMaier opened the pull request:
This is an implementation for easier writing of applications tests:

The required functionality is:

- simple implementation of tests of applications with their typical configurations
- tests of application outputs (e.g., comparison with reference data files)
- ability to use similar applications / configurations for the tests of the simtools-prod containers.

#### test_application_from_config.py

Test function which reads all yaml files in the `tests/integration_tests/config/` and runs the test. This module is very general and does not have any knowledge about the applications, with the exception that the exception is that all applications can be configures with `--config <configuration file`).

Additional points:

- output is written to "tmp_test_directory"/tool-name/test-name (to avoid that tests overwrite results from other tests)
- use Path.glob instead glob.glob
- added that all applications are called the following way:
   - simtools-<too-name> --help (disabled, as some tools fail; needs to be fixed)
   - simtools-<tool-name> (no command line parameters; disabled, as some tools fail on this)
   - simtools-<tool-name> --version
- moved tolerance into a yaml file.

#### config files

In `tests/integration_tests/config/` , they include:

- application configuration expected as SIM_PIPE:CONFIGURATION (this is an existing functionality)
- SIM_PIPE:APPLICATION - the application name (needed to run the tests)
- SIM_PIPE:INTEGRATION_TEST - allows to configure tests or give e.g., reference data files

To run the test, do:

```
pytest  -s --no-cov tests/integration_tests/test_application_from_config.py
```

- runs over the two config files available (should be announced as two tests, needs to be fixed)
- for `print_array_elements_utm_to_corsika.yml`, it compares the output astropy table to a reference table in the test/resources directory; it tests for equivalent column names and approx the same values.

VictorBarbosaMartins commented:
Is this ready for review? It is still marked as draft but you requested a review. this is confusing. Also #727 is marked for Tobias to review but is still a draft.

VictorBarbosaMartins commented on tests/resources/telescope_positions-North-corsika.ecsv:
```diff
@@ -0,0 +1,84 @@
+# %ECSV 1.0
```
We got rid of such a file a couple of months ago in order not to duplicate the telescope list in the DB. I suggest we keep either one of them but not both

GernotMaier replied:
I think the main point is to be consistent:  this is a test file for `print_array_elements`, where we convert `./tests/resources/telescope_positions-North-utm.ecsv` and test it with `telescope_positions-North-corsika.ecsv`.  We shouldn't have one file in the test directory and the other one in the DB.

I think to remember we removed it as it wasn't used in the tests, is this correct?

VictorBarbosaMartins replied:
I checked the DB and apparently we have already deleted these telescope position files, so it's all good.

VictorBarbosaMartins commented on tests/integration_tests/config/print_array_elements_utm_to_corsika.yml:
```diff
@@ -0,0 +1,14 @@
+  CONFIGURATION:
+    INPUT: ./tests/resources/telescope_positions-North-utm.ecsv
+    EXPORT: corsika
+    USE_CORSIKA_TELESCOPE_HEIGHT: True
```
my only worry with this approach is that the variables names are written in uppercase here, while in the applications they are lower case. That means, we would have to remember to always use the `-i` option from `grep` such that the search is case insensitive. See below
https://stackoverflow.com/questions/48492422/how-to-grep-for-case-insensitive-string-in-a-file

VictorBarbosaMartins replied:
I agree it is better to follow CTAO but I think it would be a good idea to include into the developer guidelines some info about this new way of testing functions, including also what I pointed out above. Of course that can be done later in another PR

orelgueta replied:
As mentioned in the call, it doesn't actually improve readability. 

Regarding following how CTAO generally writes yaml files, I am not sure the capitalisation is well motivated or that it is actually planned to be implemented that way. At least the yaml files which control ctapipe processing are not defined this way (see here for example - https://gitlab.cta-observatory.org/cta-computing/dpps/datapipe/pipeline-configurations/-/tree/main/ctapipe/v0.19.0/v3?ref_type=heads)

However, as I said in the call, my objection is limited to complaining, nothing more.

GernotMaier commented:
@orelgueta , @VictorBarbosaMartins - I will need your help on this tomorrow.

The tests related to simulation production fail since the merge-main-to-this-branch today. Something odd, as the tests on main with the old test_application.py are working. Do you know out of the top of your head what the issue could be?

(also need to understand what path parameters are not set correctly, as I still get files like run000002_gamma_za020deg_azm000deg_North_TestLayout_test-production.log_hist.tar.gz into main simtools directory. Not sure I understand the difference between `output_path` and `data_directory`)

orelgueta commented:
I fixed the issue and now tests are running without failing and without writing anything locally.
There were two issues,
- As part of a separate PR (#742), I added two parameters to the DB which must be added to the exclusion list. I don't know if you ran into that issue because I think I did it after you stopped working (I did it at night on purpose). However, this would have caused the tests to fail.
- The output and data directories were set to somewhat strange values:
  - The output directory set in the config file was basically overwritten in the code instead of being joined to the temp test directory. 
  - If the output directory was not set in the configuration dictionary, the default value from the argparser is taken, which is the local simtools-output directory (that is why you got files in the local directory).
  - The data directory had the same issue, only in this case `/data` was joined to it as well, which was not expected anywhere. 
  - One of the test had an `OUTPUT_FILE` set to a local file.

I modified the code a bit to deal with the issues above as best as I understand. I am not sure you would agree with these changes, so please review carefully.

GernotMaier commented:
This is ready for review now. (@VictorBarbosaMartins , @tobiaskleiner )

Important to note:

- includes in the config directory the same integration tests with almost exactly the same naming as in the 'old' test_application.py
- addition to configurator.py: set required=False for all parameter if a config file is given (otherwise you can't run the application - we discussed this)
- there are three TODO's, which will be fixed in an upcoming PR (working on it already), so please ignore them.
- validate_application_output needs (as discussed) some extensions. This will be done in future if additional tests added
- minor changes to applications (e.g., command line parameters should all be lower case)

Only think left is updating the documentation. I let you already start with the review while I am doing this.

GernotMaier commented:
No updates to documentation is required - the current text is fine without explicitly discussing `test_application`. So enjoy the review.

VictorBarbosaMartins commented:
Hi @GernotMaier. @tobiaskleiner and I found an error that was not allowing us to run the applications in the command line without the config argument. That was the case for all applications but here is an example. We ran:
```
simtools-generate-corsika-histograms --iact_file ./tests/resources/tel_output_10GeV-2-gamma-20deg-CTAO-South.corsikaio --pdf --output_path .
```
and got
```
Traceback (most recent call last):
  File "/workdir/env/bin/simtools-generate-corsika-histograms", line 8, in <module>
    sys.exit(main())
  File "/workdir/external/simtools/simtools/applications/generate_corsika_histograms.py", line 415, in main
    args_dict, _ = _parse(label, description)
  File "/workdir/external/simtools/simtools/applications/generate_corsika_histograms.py", line 271, in _parse
    config_parser, _ = config.initialize(db_config=False, paths=True)
  File "/workdir/external/simtools/simtools/configuration/configurator.py", line 156, in initialize
    self._fill_from_config_file(self.config.get("config", None))
  File "/workdir/external/simtools/simtools/configuration/configurator.py", line 294, in _fill_from_config_file
    _config_dict = gen.remove_substring_recursively_from_dict(_config_dict, substring="\n")
  File "/workdir/external/simtools/simtools/utils/general.py", line 806, in remove_substring_recursively_from_dict
    for key, value in data_dict.items():
AttributeError: 'NoneType' object has no attribute 'items'
```

We are committing one change that fixes it. We now only fill the config attribute if the config parameter is there. Please have a look if you agree with the proposed solution so that we continue with the review. If you prefer to go for a different solution, please undo the commit. Thanks!

VictorBarbosaMartins commented:
@GernotMaier, we are now looking at the new function that removes the "\n" but we could not easily find where this would be useful. Do you know in which tests exactly yaml.load was adding "\n" to a value? We are investigating an alternative approach

GernotMaier commented:
> Hi @GernotMaier. @tobiaskleiner and I found an error that was not allowing us to run the applications in the command line without the config argument. That was the case for all applications but here is an example. We ran:
> 
> ```
> simtools-generate-corsika-histograms --iact_file ./tests/resources/tel_output_10GeV-2-gamma-20deg-CTAO-South.corsikaio --pdf --output_path .
> ```
> 
> and got
> 
> ```
> Traceback (most recent call last):
>   File "/workdir/env/bin/simtools-generate-corsika-histograms", line 8, in <module>
>     sys.exit(main())
>   File "/workdir/external/simtools/simtools/applications/generate_corsika_histograms.py", line 415, in main
>     args_dict, _ = _parse(label, description)
>   File "/workdir/external/simtools/simtools/applications/generate_corsika_histograms.py", line 271, in _parse
>     config_parser, _ = config.initialize(db_config=False, paths=True)
>   File "/workdir/external/simtools/simtools/configuration/configurator.py", line 156, in initialize
>     self._fill_from_config_file(self.config.get("config", None))
>   File "/workdir/external/simtools/simtools/configuration/configurator.py", line 294, in _fill_from_config_file
>     _config_dict = gen.remove_substring_recursively_from_dict(_config_dict, substring="\n")
>   File "/workdir/external/simtools/simtools/utils/general.py", line 806, in remove_substring_recursively_from_dict
>     for key, value in data_dict.items():
> AttributeError: 'NoneType' object has no attribute 'items'
> ```
> 
> We are committing one change that fixes it. We now only fill the config attribute if the config parameter is there. Please have a look if you agree with the proposed solution so that we continue with the review. If you prefer to go for a different solution, please undo the commit. Thanks!

Thanks for catching this one - I've reverted your changes and add the right error for the Exception statements (which passes) in _fill_from_config_file (add the addition of `gen.remove_substring_recursively_from_dict()` changed the error type raised. Added also a unit test for this case.

VictorBarbosaMartins reviewed (changes requested):
Thanks for this nice implementation. I have a several of minor comments and two major comments:
- By running the integration tests in main I get 72 tests, while running in this branch I get 69 tests. Apparently three tests have gone missing?
- (also find in the individual comments) I am not convinced with the approach of saving to temp_config.yml before running the application test with "--config". In summary, the code reads and appends all the configuration dicts from the config directory, but then it writes the configuration again to temp_config to run the application with config from file. I was expecting it would run directly from the config files in the config directory. If that is not straightforward to implement, at least running the tests from the read dictionary would be more direct than writing a temporary file in between. We can discuss this tomorrow if you like.

GernotMaier commented on tests/integration_tests/config/plot_array_layout_one_file.yml:
```diff
@@ -0,0 +1,6 @@
+  TEST_NAME: one_file
+  CONFIGURATION:
+    TELESCOPE_LIST: data/layout/telescope_positions-North-TestLayout.ecsv
+    OUTPUT_PATH: simtools-tests
```
Not all application write to disk (e.g., add_file_to_db.py and validate_file_using_schema.py). So yes, this is needed.

GernotMaier commented on tests/integration_tests/config/sim_showers_for_trigger_rates_run.yml:
```diff
@@ -0,0 +1,14 @@
+    NEVENTS: 10000
+    TEST: true
+    SUBMIT_COMMAND: local
+    DATA_DIRECTORY: simtools-output
+    OUTPUT_PATH: simtools-output
+    DATA_DIRECTORY: simtools-output
```
Fixed.

GernotMaier commented on tests/integration_tests/config/plot_array_layout_layout_name.yml:
I systematically build the names from the old application tests with 'application name' + 'test name'. This sometimes sounds weird. Anyway, I changed this to `plot_array_layout_by_name.yml`

GernotMaier commented on tests/integration_tests/test_applications_from_config.py:
```diff
@@ -0,0 +1,205 @@
+    Return the command to run the application with the given config file.
+
+    """
+    if app.find("simtools-") < 0:
+        cmd = "python simtools/applications/" + app + ".py"
```
Yes, this should stay - two lines which increase flexibility.

GernotMaier commented on tests/integration_tests/test_applications_from_config.py:
```diff
@@ -0,0 +1,205 @@
+    # list of all applications
+    _applications = list(set(item["APPLICATION"] for item in configs if "APPLICATION" in item))
+    for _app in _applications:
+        # add for all applications "--help" call
```
Don't worry about these lines, I am working on them.

GernotMaier commented on tests/integration_tests/test_applications_from_config.py:
```diff
@@ -0,0 +1,205 @@
+    _applications = list(set(item["APPLICATION"] for item in configs if "APPLICATION" in item))
+    for _app in _applications:
+        # add for all applications "--help" call
+        # TODO - disabled, as not all application have help implemented
+        # configs.append(
+        # {"APPLICATION": _app, "TEST_NAME": "auto-help", "CONFIGURATION": {"HELP": True}})
+        # add for all applications "--version" call
+        # configs.append(
+        #    {"APPLICATION": _app, "TEST_NAME": "auto-version", "CONFIGURATION": {"VERSION": True}}
+        # )
+        # TODO - disabled, as not all applications can be called without command line parameters
+        # add for all applications call without config file
+        # configs.append({"APPLICATION": _app, "TEST_NAME": "auto-no_config"})
+        logger.info("Missing implementations of help, versions, no command line parameter")
```
I am working on it - an issue is really not needed.

GernotMaier commented on tests/integration_tests/config/print_array_elements_utm_to_corsika.yml:
```diff
@@ -0,0 +1,14 @@
+    OUTPUT_FILE: test.ecsv
+    SELECT_ASSETS: ['LSTN', 'MSTN']
+  INTEGRATION_TESTS:
+    - REFERENCE_OUTPUT_FILE: |
+        ./tests/resources/telescope_positions-North-corsika.ecsv
```
I think we should discuss the structure of this directory when we have too many files, which is not the case yet. Note that in this case, the output can also be an input (as we can convert in all directions).

VictorBarbosaMartins replied:
Then I would suggest to include the test in the other direction as well (corsika to utm) which is missing. I know there are more tests missing and they will be included later but by ignoring it, we don't keep track of what is missing in the end.

GernotMaier commented on tests/integration_tests/test_applications_from_config.py:
```diff
@@ -0,0 +1,205 @@
+    return configs
+
+
+def compare_ecsv_files(file1, file2, tolerance=1.0e-5):
```
Many comparisons / validation steps are missing, as we have discussed yesterday. Mire tests should be added in the future, let's keep this pull request on reasonable size.

VictorBarbosaMartins replied:
sure, but I think adding two lines to assert that both metadatas are actually the same wouldn't be a big deal in my opinion. I can do it if you prefer

GernotMaier commented on tests/integration_tests/test_applications_from_config.py:
```diff
@@ -0,0 +1,205 @@
+                integration_test.get("TOLERANCE", 1.0e-5),
+            )
+
+    return 0
```
Reason was that the previous function calls system, in which "OK" is equivalent to a 0 as return pattern.

VictorBarbosaMartins replied:
Ok, I see now. This is actually helpful to check if the validation was successful.

GernotMaier commented on tests/integration_tests/test_applications_from_config.py:
```diff
@@ -0,0 +1,205 @@
+def get_tmp_config_file(config, output_path):
+    """
+    Write a temporary config file for the application to be tested.
+    Change output path and file name to values suitable for tests.
```
Fixed this.

GernotMaier commented on tests/integration_tests/test_applications_from_config.py:
```diff
@@ -0,0 +1,205 @@
+    return 0
+
+
+def get_tmp_config_file(config, output_path):
```
Yes and no - it gets the path + file name of the temporary configuration file. I changed it to `write_and_get_tmp_config_file`, but it really doesn't help.

VictorBarbosaMartins replied:
Then I suggest dividing the function into two. Note that if one wants to get the file, only the very first line of the function out of 11 is used.

GernotMaier replied:
I really don't think this is needed - this is a small function and it leaks very readable to me.

GernotMaier commented on tests/integration_tests/test_applications_from_config.py:
```diff
@@ -0,0 +1,205 @@
+    with open(tmp_config_file, "w", encoding="utf-8") as file:
+        yaml.safe_dump(config, file, sort_keys=False)
+
+    return tmp_config_file
```
I don't think this is a performance issue at all, writing a single file is not an issue giving that the applications are writing 10s to 100s of files. 

Writing a temporary file is in this case the most efficient way to this. It also allows easy debugging (as one can repeat the test locally using the config file).

VictorBarbosaMartins replied:
Ok. Indeed it is not a performance issue but more like structural.

For the record, I checked the time for running the integration tests from main and from the branch and, in CPU time, the main is slower:

Main:
```
real	6m46.611s
user	21m0.854s
sys	1m45.397s
```

Branch:
```
real	7m6.134s
user	19m44.759s
sys	1m39.729s
```

GernotMaier commented on tests/integration_tests/test_applications_from_config.py:
```diff
@@ -0,0 +1,205 @@
+    get_list_of_test_configurations(),
+    ids=get_list_of_test_configurations(get_test_names=True),
+)
+def test_applications_from_config(tmp_test_directory, config, monkeypatch):
+    """
+    Test all applications from config files found in the config directory.
+    Test output is written to a temporary directory and tested.
+
+    """
```
These are not parameters used by any user when calling the function, but value filled by the pytest fixtures. This is why we don't document them (here and in all other test functions).

VictorBarbosaMartins replied:
Sure, but this is the major function where all integration tests run. The documentation is for ourselves, the developers, and it helps us understanding the function before looking into the code.

GernotMaier replied:
But we nowhere calling this function, these are all pytest fixtures as defined above the function call. The fixture all have proper docstring, I really don't see what it adds here.

VictorBarbosaMartins replied:
I pushed a commit with the docstring that I meant. this is usual and nothing new to the project

GernotMaier commented on tests/integration_tests/test_applications_from_config.py:
```diff
@@ -0,0 +1,205 @@
+        logger.error(f"No application defined in config file {config}.")
+        raise exc
+
+    tmp_output_path.mkdir(parents=True, exist_ok=True)
```
- tests should be independent of each other - having each writing in its own directory ensures that one test does not use by accident files from another one
- try and run the tests locally and debug something - having them in separate directories is very helpful

VictorBarbosaMartins replied:
Ok, agree with the first point.

GernotMaier replied:
This is resolved then?

VictorBarbosaMartins replied:
yes, the conversation is already resolved

GernotMaier commented:
@tobiaskleiner @VictorBarbosaMartins - thanks for the review!

I think I've addressed all your points, please let me know if something missing.

> Thanks for this nice implementation. I have a several of minor comments and two major comments:
> 
>     * By running the integration tests in main I get 72 tests, while running in this branch I get 69 tests. Apparently three tests have gone missing?

These are the tests which either call the application with arguments, or with `--help` only. They don't need to be written using configuration files.

> 
>     * (also find in the individual comments) I am not convinced with the approach of saving to temp_config.yml before running the application test with "--config". In summary, the code reads and appends all the configuration dicts from the config directory, but then it writes the configuration again to temp_config to run the application with config from file. I was expecting it would run directly from the config files in the config directory. If that is not straightforward to implement, at least running the tests from the read dictionary would be more direct than writing a temporary file in between. We can discuss this tomorrow if you like.

As mentioned above, this is the most straightforward way to do it, given that we modify the output paths (this is done as we can use the config files as well in other cases, e.g., the simtools-prod tests), and a much more obvious and easier way of following up what is happening / debugging.

VictorBarbosaMartins commented:
> @tobiaskleiner @VictorBarbosaMartins - thanks for the review!
> 
> I think I've addressed all your points, please let me know if something missing.
> 
> > Thanks for this nice implementation. I have a several of minor comments and two major comments:
> > ```
> > * By running the integration tests in main I get 72 tests, while running in this branch I get 69 tests. Apparently three tests have gone missing?
> > ```
> 
> These are the tests which either call the application with arguments, or with `--help` only. They don't need to be written using configuration files.
> 
> > ```
> > * (also find in the individual comments) I am not convinced with the approach of saving to temp_config.yml before running the application test with "--config". In summary, the code reads and appends all the configuration dicts from the config directory, but then it writes the configuration again to temp_config to run the application with config from file. I was expecting it would run directly from the config files in the config directory. If that is not straightforward to implement, at least running the tests from the read dictionary would be more direct than writing a temporary file in between. We can discuss this tomorrow if you like.
> > ```
> 
> As mentioned above, this is the most straightforward way to do it, given that we modify the output paths (this is done as we can use the config files as well in other cases, e.g., the simtools-prod tests), and a much more obvious and easier way of following up what is happening / debugging.

Hi @GernotMaier. There are still some open points.
 
I do not agree that this is the most straightforward way to do it. The code is reading, appending to a list all dicts and writing almost the same files as it had read before to then finally run the test by reading from file again. One could very well repeat the test by using the configuration file from the config directory, which does not change, without the need of the temporary file. But if you prefer to keep it like this, it is of course your call. I still have two more points in this regard after your answers:
- Why would one like to have a different output path for the integration tests for simtools-prod such that the use of a temporary config file is necessary for changing the output path?
- Regard debugging the integration test. It is for me easier to debug by looking at the print outs, i.e., printing out the dictionary with the config file, and not digging into subdirectories. But that's also just my simple opinion.


In regard to the missing 3 tests, ok. Let's just not forget them later on.

GernotMaier commented:
Thanks again for the comments. 

I think we have discussed in the chat why those temporary files are important and this is resolved.
The other points have been addressed, or are in my opinion a style question - please let me know if I have overlooked something.

GernotMaier commented:
Thank you for the review!
